File: src/level_manager.cpp
#include "level_manager.h"
#include <iostream>
#include <vector>
#include <string>
#include <iomanip>
#include <algorithm>

LevelManager::LevelManager() : totalScore(0) {
    initializeLevels();
}

void LevelManager::initializeLevels() {
    // 确保清空现有关卡
    levels.clear();
    
    // 添加16个关卡
    levels = {
        // 基础关卡 (1-6)
        {1, "有理数运算", "掌握基本的有理数加减乘除", LEVEL_UNLOCKED, 0, 0, 0},
        {2, "一元一次方程", "解一元一次方程的基础练习", LEVEL_LOCKED, 0, 0, 30},
        {3, "因式分解", "多项式因式分解技巧", LEVEL_LOCKED, 0, 0, 60},
        {4, "小数运算", "十以内小数加减乘除", LEVEL_LOCKED, 0, 0, 90},
        {5, "百以内小数", "百以内小数加减乘除", LEVEL_LOCKED, 0, 0, 120},
        {6, "复杂小数运算", "多步复杂小数运算", LEVEL_LOCKED, 0, 0, 150},
        
        // 新增关卡 (7-12)
        {7, "幂函数计算", "掌握幂函数的基本运算", LEVEL_LOCKED, 0, 0, 180},
        {8, "分数幂运算", "分数幂的运算与应用", LEVEL_LOCKED, 0, 0, 210},
        {9, "平方根计算", "平方根的运算与应用", LEVEL_LOCKED, 0, 0, 240},
        {10, "三角函数基础", "特殊角的三角函数值", LEVEL_LOCKED, 0, 0, 270},
        {11, "三角方程应用", "简单三角方程求解", LEVEL_LOCKED, 0, 0, 300},
        {12, "三角恒等式", "复合三角恒等式与不等式", LEVEL_LOCKED, 0, 0, 330},
        
        // 综合关卡 (13-16)
        {13, "综合关卡一", "综合前面所有知识点的挑战", LEVEL_LOCKED, 0, 0, 360},
        {14, "常数计算", "π值、e值和勾股定理的应用", LEVEL_LOCKED, 0, 0, 390},
        {15, "对数与阶乘", "对数和阶乘的计算", LEVEL_LOCKED, 0, 0, 420},
        {16, "综合关卡二", "终极综合挑战", LEVEL_LOCKED, 0, 0, 450}
    };
    
    // 调试输出，确认关卡数量
    std::cout << "DEBUG: 初始化了 " << levels.size() << " 个关卡" << std::endl;
}

Level LevelManager::getLevel(int levelId) {
    if (levelId < 1 || levelId > levels.size()) {
        std::cout << "DEBUG: 请求的关卡ID " << levelId << " 超出范围 (1-" << levels.size() << ")" << std::endl;
        return Level{0, "", "", LEVEL_LOCKED, 0, 0, 0};
    }
    
    return levels[levelId - 1];
}

std::vector<Level> LevelManager::getAllLevels() {
    std::cout << "DEBUG: getAllLevels() 返回 " << levels.size() << " 个关卡" << std::endl;
    return levels;
}

bool LevelManager::unlockLevel(int levelId) {
    if (levelId < 1 || levelId > levels.size()) {
        std::cout << "DEBUG: 尝试解锁无效关卡ID " << levelId << std::endl;
        return false;
    }
    
    Level& level = levels[levelId - 1];
    if (level.status == LEVEL_LOCKED && totalScore >= level.requiredScore) {
        level.status = LEVEL_UNLOCKED;
        std::cout << "DEBUG: 已解锁关卡 " << levelId << std::endl;
        return true;
    }
    
    return false;
}

bool LevelManager::completeLevel(int levelId, int score, int stars) {
    if (levelId < 1 || levelId > levels.size()) {
        std::cout << "DEBUG: 尝试完成无效关卡ID " << levelId << std::endl;
        return false;
    }
    
    Level& level = levels[levelId - 1];
    if (level.status == LEVEL_UNLOCKED) {
        level.status = LEVEL_COMPLETED;
        level.score = score;
        level.stars = stars;
        totalScore += score;
        
        std::cout << "DEBUG: 已完成关卡 " << levelId << ", 得分: " << score << ", 星级: " << stars << std::endl;
        
        // 尝试解锁下一关
        if (levelId < levels.size()) {
            unlockLevel(levelId + 1);
        }
        
        return true;
    }
    
    return false;
}

int LevelManager::getTotalScore() const {
    return totalScore;
}

void LevelManager::addScore(int score) {
    totalScore += score;
    std::cout << "DEBUG: 添加分数 " << score << ", 当前总分: " << totalScore << std::endl;
}

void LevelManager::displayLevelSelection() {
    std::cout << "\n=== 选择闯关关卡 ===\n";
    std::cout << "当前总分: " << totalScore << "\n";
    std::cout << "DEBUG: 显示关卡选择, 共有 " << levels.size() << " 个关卡\n\n";
    
    // 分页显示关卡
    int currentPage = 1;
    const int levelsPerPage = 8;
    int totalPages = (levels.size() + levelsPerPage - 1) / levelsPerPage;
    
    while (true) {
        std::cout << "--- 第 " << currentPage << " 页/共 " << totalPages << " 页 ---\n";
        
        int startIndex = (currentPage - 1) * levelsPerPage;
        int endIndex = std::min(startIndex + levelsPerPage - 1, static_cast<int>(levels.size()) - 1);
        
        for (int i = startIndex; i <= endIndex; i++) {
            const auto& level = levels[i];
            std::cout << level.id << ". " << level.title;
            
            switch (level.status) {
                case LEVEL_LOCKED:
                    std::cout << " [锁定 - 需要 " << level.requiredScore << " 分]";
                    break;
                case LEVEL_UNLOCKED:
                    std::cout << " [进行中]";
                    break;
                case LEVEL_COMPLETED:
                    std::cout << " [已完成]";
                    break;
            }
            
            std::cout << " ";
            for (int j = 0; j < level.stars; j++) {
                std::cout << "★";
            }
            std::cout << std::endl;
        }
        
        // 显示导航选项
        std::cout << "\n导航选项:\n";
        if (currentPage > 1) {
            std::cout << "P. 上一页\n";
        }
        if (currentPage < totalPages) {
            std::cout << "N. 下一页\n";
        }
        std::cout << "0. 返回主菜单\n";
        std::cout << "请选择关卡编号或操作: ";
        
        std::string input;
        std::getline(std::cin, input);
        
        // 处理导航选项
        if (input == "P" || input == "p") {
            if (currentPage > 1) currentPage--;
            continue;
        } else if (input == "N" || input == "n") {
            if (currentPage < totalPages) currentPage++;
            continue;
        } else if (input == "0") {
            return;
        }
        
        // 尝试将输入转换为关卡编号
        try {
            int choice = std::stoi(input);
            if (choice >= 1 && choice <= levels.size()) {
                std::cout << "已选择关卡 " << choice << std::endl;
                // 这里应该调用关卡开始函数
                return;
            } else {
                std::cout << "无效的关卡编号! 请输入 1-" << levels.size() << " 之间的数字。\n";
            }
        } catch (...) {
            std::cout << "无效输入! 请输入数字或操作指令。\n";
        }
        
        std::cout << "按回车键继续...";
        std::cin.get();
    }
}

void LevelManager::displayLevelProgress() {
    std::cout << "\n=== 我的闯关进度 ===\n";
    std::cout << "当前总分: " << totalScore << "\n";
    std::cout << "DEBUG: 显示关卡进度, 共有 " << levels.size() << " 个关卡\n\n";
    
    // 表头
    std::cout << std::left << std::setw(4) << "ID" 
              << std::setw(20) << "关卡名称" 
              << std::setw(10) << "状态" 
              << std::setw(8) << "星级" 
              << std::setw(10) << "分数" 
              << std::setw(15) << "解锁要求" 
              << std::endl;
    std::cout << std::string(67, '-') << std::endl;
    
    // 显示所有关卡
    for (const auto& level : levels) {
        std::cout << std::left << std::setw(4) << level.id;
        std::cout << std::setw(20) << (level.title.length() > 18 ? level.title.substr(0, 15) + "..." : level.title);
        
        // 状态
        std::string statusStr;
        switch (level.status) {
            case LEVEL_LOCKED: statusStr = "锁定"; break;
            case LEVEL_UNLOCKED: statusStr = "进行中"; break;
            case LEVEL_COMPLETED: statusStr = "已完成"; break;
        }
        std::cout << std::setw(10) << statusStr;
        
        // 星级
        std::string starsStr;
        for (int i = 0; i < level.stars; i++) {
            starsStr += "★";
        }
        if (starsStr.empty()) starsStr = "无";
        std::cout << std::setw(8) << starsStr;
        
        // 分数
        std::cout << std::setw(10) << (level.status == LEVEL_COMPLETED ? std::to_string(level.score) + "/100" : "-");
        
        // 解锁要求
        std::cout << std::setw(15) << (level.status == LEVEL_LOCKED ? std::to_string(level.requiredScore) + "分" : "已解锁");
        
        std::cout << std::endl;
    }
    
    // 统计信息
    int completed = 0;
    int totalStars = 0;
    for (const auto& level : levels) {
        if (level.status == LEVEL_COMPLETED) {
            completed++;
            totalStars += level.stars;
        }
    }
    
    std::cout << "\n统计: " << completed << "/" << levels.size() << " 关卡已完成, ";
    std::cout << "总共获得 " << totalStars << "/" << (levels.size() * 3) << " 颗星\n";
    
    std::cout << "\n按回车键返回主菜单...";
    std::cin.get();
}

int LevelManager::getTotalLevels() const {
    std::cout << "DEBUG: getTotalLevels() 返回 " << levels.size() << std::endl;
    return levels.size();
}

// 添加一个调试函数，用于检查关卡状态
void LevelManager::debugLevels() {
    std::cout << "DEBUG: 当前关卡状态:\n";
    for (const auto& level : levels) {
        std::cout << "关卡 " << level.id << ": " << level.title 
                  << " (状态: " << level.status 
                  << ", 分数: " << level.score 
                  << ", 星级: " << level.stars 
                  << ", 解锁要求: " << level.requiredScore << ")\n";
    }
}

File: level_manager.js
// level_manager.js
// 移植自 C++ level_manager.cpp

export const LevelStatus = {
  LEVEL_LOCKED: 'LEVEL_LOCKED',
  LEVEL_UNLOCKED: 'LEVEL_UNLOCKED',
  LEVEL_COMPLETED: 'LEVEL_COMPLETED'
};

export class Level {
  // unlockRule: optional object { type: 'prevStars'|'prevScore', value: number }
  constructor(id, title, description, status = LevelStatus.LEVEL_LOCKED, score = 0, stars = 0, requiredScore = 0, unlockRule = null) {
    this.id = id;
    this.title = title;
    this.description = description;
    this.status = status;
    this.score = score;
    this.stars = stars;
    this.requiredScore = requiredScore;
    this.unlockRule = unlockRule; // new
  }
}

export class LevelManager {
  constructor() {
    this.levels = [];
    this.totalScore = 0;
    this.initializeLevels();
  }

  initializeLevels() {
    // By default, use unlockRule that requires previous level to have at least 1 star.
    // This can be customized per level by modifying unlockRule.
    this.levels = [
      new Level(1, '有理数运算', '掌握基本的有理数加减乘除', LevelStatus.LEVEL_UNLOCKED, 0, 0, 0, null),
      new Level(2, '一元一次方程', '解一元一次方程的基础练习', LevelStatus.LEVEL_LOCKED, 0, 0, 30, { type: 'prevStars', value: 1 }),
      new Level(3, '因式分解', '多项式因式分解技巧', LevelStatus.LEVEL_LOCKED, 0, 0, 60, { type: 'prevStars', value: 1 }),
      new Level(4, '小数运算', '十以内小数加减乘除', LevelStatus.LEVEL_LOCKED, 0, 0, 90, { type: 'prevStars', value: 1 }),
      new Level(5, '百以内小数', '百以内小数加减乘除', LevelStatus.LEVEL_LOCKED, 0, 0, 120, { type: 'prevStars', value: 1 }),
      new Level(6, '复杂小数运算', '多步复杂小数运算', LevelStatus.LEVEL_LOCKED, 0, 0, 150, { type: 'prevStars', value: 1 }),
      new Level(7, '幂函数计算', '掌握幂函数的基本运算', LevelStatus.LEVEL_LOCKED, 0, 0, 180, { type: 'prevStars', value: 1 }),
      new Level(8, '分数幂运算', '分数幂的运算与应用', LevelStatus.LEVEL_LOCKED, 0, 0, 210, { type: 'prevStars', value: 1 }),
      new Level(9, '平方根计算', '平方根的运算与应用', LevelStatus.LEVEL_LOCKED, 0, 0, 240, { type: 'prevStars', value: 1 }),
      new Level(10, '三角函数基础', '特殊角的三角函数值', LevelStatus.LEVEL_LOCKED, 0, 0, 270, { type: 'prevStars', value: 1 }),
      new Level(11, '三角方程应用', '简单三角方程求解', LevelStatus.LEVEL_LOCKED, 0, 0, 300, { type: 'prevStars', value: 1 }),
      new Level(12, '三角恒等式', '复合三角恒等式与不等式', LevelStatus.LEVEL_LOCKED, 0, 0, 330, { type: 'prevStars', value: 1 }),
      new Level(13, '综合关卡一', '综合前面所有知识点的挑战', LevelStatus.LEVEL_LOCKED, 0, 0, 360, { type: 'prevStars', value: 1 }),
      new Level(14, '常数计算', 'π值、e值和勾股定理的应用', LevelStatus.LEVEL_LOCKED, 0, 0, 390, { type: 'prevStars', value: 1 }),
      new Level(15, '对数与阶乘', '对数和阶乘的计算', LevelStatus.LEVEL_LOCKED, 0, 0, 420, { type: 'prevStars', value: 1 }),
      new Level(16, '综合关卡二', '终极综合挑战', LevelStatus.LEVEL_LOCKED, 0, 0, 450, { type: 'prevStars', value: 1 })
    ];
    this.totalScore = 0;
  }

  getLevel(levelId) {
    if (levelId < 1 || levelId > this.levels.length) return null;
    return this.levels[levelId - 1];
  }

  getAllLevels() {
    return this.levels.slice();
  }

  unlockLevel(levelId) {
    if (levelId < 1 || levelId > this.levels.length) return false;
    const level = this.levels[levelId - 1];
    if (level.status !== LevelStatus.LEVEL_LOCKED) return false;
    // If unlockRule exists, evaluate based on that first
    if (level.unlockRule && typeof level.unlockRule === 'object') {
      const rule = level.unlockRule;
      if (rule.type === 'prevStars') {
        const prev = this.getLevel(levelId - 1);
        const prevStars = prev ? (prev.stars || 0) : 0;
        console.log(`[level_manager] unlockLevel check prevStars for level ${levelId}: prevStars=${prevStars}, required=${rule.value}`);
        if (prev && prevStars >= (rule.value || 0)) {
          level.status = LevelStatus.LEVEL_UNLOCKED;
          console.log(`[level_manager] level ${levelId} unlocked by prevStars`);
          return true;
        }
        console.log(`[level_manager] level ${levelId} remains locked by prevStars`);
        return false;
      }
      if (rule.type === 'prevScore') {
        const prev = this.getLevel(levelId - 1);
        const prevScore = prev ? (prev.score || 0) : 0;
        console.log(`[level_manager] unlockLevel check prevScore for level ${levelId}: prevScore=${prevScore}, required=${rule.value}`);
        if (prev && prevScore >= (rule.value || 0)) {
          level.status = LevelStatus.LEVEL_UNLOCKED;
          console.log(`[level_manager] level ${levelId} unlocked by prevScore`);
          return true;
        }
        console.log(`[level_manager] level ${levelId} remains locked by prevScore`);
        return false;
      }
    }
    // Fallback: unlock by totalScore threshold
    if (this.totalScore >= level.requiredScore) {
      level.status = LevelStatus.LEVEL_UNLOCKED;
      return true;
    }
    return false;
  }

  completeLevel(levelId, score, stars) {
    if (levelId < 1 || levelId > this.levels.length) return false;
    const level = this.levels[levelId - 1];
    // Allow completing a level that is unlocked or already completed (replay case).
    if (level.status === LevelStatus.LEVEL_UNLOCKED || level.status === LevelStatus.LEVEL_COMPLETED) {
      console.log(`[level_manager] completeLevel called for ${levelId} score=${score} stars=${stars} (prevStatus=${level.status}, prevScore=${level.score}, prevStars=${level.stars})`);
      // Mark as completed
      const prevScore = (level.status === LevelStatus.LEVEL_COMPLETED && typeof level.score === 'number') ? level.score : 0;
      // Only update stored score and totalScore if new score is higher than previous best
      if (score > prevScore) {
        // apply difference to totalScore
        this.totalScore += (score - prevScore);
        level.score = score;
        // keep the best star count (or prefer new if higher)
        level.stars = Math.max(level.stars || 0, stars || 0);
      } else {
        // do not lower stored score; but still consider increasing stars if provided
        level.stars = Math.max(level.stars || 0, stars || 0);
      }
      level.status = LevelStatus.LEVEL_COMPLETED;
      console.log(`[level_manager] level ${levelId} marked completed: newScore=${level.score}, newStars=${level.stars}, totalScore=${this.totalScore}`);
      // ensure next level is unlocked if possible
      if (levelId < this.levels.length) this.unlockLevel(levelId + 1);
      return true;
    }
    return false;
  }

  getTotalScore() {
    return this.totalScore;
  }

  addScore(score) {
    this.totalScore += score;
  }

  displayLevelSelection() {
    console.log('\n=== 选择闯关关卡 ===\n总分: ' + this.totalScore + '\n');
    const levelsPerPage = 8;
    const totalPages = Math.ceil(this.levels.length / levelsPerPage);
    for (let page = 1; page <= totalPages; page++) {
      console.log(`--- 第 ${page} 页/共 ${totalPages} 页 ---`);
      const start = (page - 1) * levelsPerPage;
      const end = Math.min(start + levelsPerPage, this.levels.length);
      for (let i = start; i < end; i++) {
        const level = this.levels[i];
        console.log(`${level.id}. ${level.title} [${level.status}] ${'★'.repeat(level.stars)}`);
      }
    }
  }

  displayLevelProgress() {
    console.log('\n=== 我的闯关进度 ===');
    console.log('当前总分: ' + this.totalScore + '\n');
    console.log('ID 关卡名称                 状态       星级  分数   解锁要求');
    console.log('-----------------------------------------------------------------');
    let completed = 0, totalStars = 0;
    for (const level of this.levels) {
      const starsStr = level.stars > 0 ? '★'.repeat(level.stars) : '无';
      const scoreStr = level.status === LevelStatus.LEVEL_COMPLETED ? `${level.score}/100` : '-';
      const unlockReq = level.status === LevelStatus.LEVEL_LOCKED ? `${level.requiredScore}分` : '已解锁';
      console.log(`${level.id.toString().padEnd(3)} ${level.title.padEnd(20)} ${level.status.padEnd(10)} ${starsStr.padEnd(4)} ${scoreStr.padEnd(6)} ${unlockReq}`);
      if (level.status === LevelStatus.LEVEL_COMPLETED) { completed++; totalStars += level.stars; }
    }
    console.log(`\n统计: ${completed}/${this.levels.length} 关卡已完成, 总共获得 ${totalStars}/${this.levels.length * 3} 颗星\n`);
  }

  getTotalLevels() {
    return this.levels.length;
  }

  debugLevels() {
    console.log('DEBUG: 当前关卡状态:');
    for (const level of this.levels) {
      console.log(`关卡 ${level.id}: ${level.title} (状态: ${level.status}, 分数: ${level.score}, 星级: ${level.stars}, 解锁要求: ${level.requiredScore})`);
    }
  }
}
